src/imodmodel/parsers.py

```diff
@@ -40,6 +43,20 @@ def _parse_from_format_str(file: BinaryIO, format_str: str) -> Tuple[Any, ...]:
     return struct.unpack(file.read(struct.size))
 
 
+def _parse_from_type_flags(file: BinaryIO, flags: int) -> Any:
+    flag_mask, flag_int, flag_float, flag_short, flag_byte = 3, 0, 1, 2, 3
+    if flags & flag_mask == flag_int:
+        return _parse_from_format_str(file, '>i')[0]
+    elif flags & flag_mask == flag_float:
+        return _parse_from_format_str(file, '>f')[0]
+    elif flags & flag_mask == flag_short:
+        return _parse_from_format_str(file, '>2h')
+    elif flags & flag_mask == flag_byte:
+        return _parse_from_format_str(file, '>4c')
+    else:
+        raise ValueError(f'Invalid flags: {flags}')
```

Comment:
this looks like witchcraft, could you explain? 🙂

Reply:
I've added some explanation to the code, but basically I'm checking for a 2-bit flag for the type of `index` and `value`, which are stored as unions in the file. The first two bits indicate the type of `index`, the following two the type of `value`. That's why I'm shifting the flags by two bits in [_parse_general_storage](https://github.com/MoritzWM/imodmodel/blob/26429c3cd1a6b5715b92a04d3cd0044719e7d794/src/imodmodel/parsers.py#L133).

There are 4 more bit flags described in the model spec:
```
Bit 4 on : "index" is not really an index
Bit 5 on : Revert to default value
Bit 6 on : Index is surface number, not contour
Bit 7 on : The change applies to only one point
```
But for parsing index and value, they are not directly relevant I suppose.

Reply:
Cool, thank you! Would you mind adding tests covering this? It's a bit complicated and I want to make sure I don't break it in the future

Reply:
Done! I added tests for [_parse_general_storage](https://github.com/MoritzWM/imodmodel/blob/dad7854104d11826c6e334c3246107b9ad6d6e67/tests/test_parsers.py#L217) and for [_parse_from_type_flags](https://github.com/MoritzWM/imodmodel/blob/dad7854104d11826c6e334c3246107b9ad6d6e67/tests/test_parsers.py#L208), and immediately discovered a small bug.